initial collator instructions (#102)
InvArch/InvArch-Node: docs/collator-instructions.md

# How to set up an InvArch Collator

Make sure you are logged in as a user with root privileges, otherwise:

``sudo su -``

Create a new service user to run your collator service:
`sudo useradd --no-create-home --shell /usr/sbin/nologin tinkernet-collator`

Install some required stuff:

``curl https://sh.rustup.rs -sSf | sh``
(choose option 1 - Proceed with installation (default))
You may need to restart your system before the next steps.
```
rustup install nightly-2021-11-02

> rustup target add wasm32-unknown-unknown --toolchain nightly-2021-11-02

Reviewer: These should not be needed as they would install the incorrect version anyway, as per the last release rustup should auto install all components for the version specified in the rust-toolchain.toml

apt install cmake git clang libclang-dev
```
Type Y to proceed.

Clone the repo:

 ``git clone git@github.com:InvArch/InvArch-Node.git``

> Type the following command, and make sure you can see "gm-chain", if not then you likely did something wrong.

Reviewer: ```suggestion
Type the following command, and make sure you can see "InvArch-Node", if not then you likely did something wrong.
```

  ``ls -la``

> Enter the repo and check out the most recent tagged release of code (https://github.com/GMorDIE/gm-chain/releases)

Reviewer: ```suggestion
Enter the repo and check out the most recent tagged release of code (https://github.com/InvArch/InvArch-Node/releases)
```

```
cd InvArch-Node

> git checkout v1.2.0

Reviewer: ```suggestion
git checkout $(git describe --tags $(git rev-list --tags --max-count=1))
```
This checks out the latest local tag

```

You need to compile the code, this will take quite a while depending on your system (30+ minutes is normal):

>  ``cargo build --release``

Reviewer: ```suggestion
 ``cargo build --release --features tinkernet``
```
Default feature is Tinkernet right now, just future-proofing these instructions

> Mid-way through compiling, you likely need to enter this command when prompted:
> 
> ``rustup target add wasm32-unknown-unknown``
> 
> Resume compiling:
> 
> ``cargo build --release``
> 

Reviewer: ```suggestion
```
This shouldn't be necessary because of the reason described in the first comment

Move the node executable to `usr/local/bin`, make it executable, and change ownership to our `tinkernet-collator` service user:
```
sudo mv ~/InvArch-Node/target/release/invarch-collator /usr/local/bin/tinkernet-collator
sudo chmod +x /usr/local/bin/tinkernet-collator
sudo chown tinkernet-collator:tinkernet-collator /usr/local/bin/tinkernet-collator
```

> Create the base-path folder, copy the gm_chain "chainspec" into it, and give it the necessary permissions & ownership:

Reviewer: ```suggestion
Create the base-path folder, copy the tinker-raw "chainspec" into it, and give it the necessary permissions & ownership:
```

```
sudo mkdir /var/lib/tinkernet
sudo cp ~/InvArch-Node/res/kusama/tinker-raw.json /var/lib/tinkernet/tinker-raw.json
sudo chown -R tinkernet-collator:tinkernet-collator /var/lib/tinkernet
```

Create a systemd service file to run your collator (and automatically restart it):

`sudo nano /etc/systemd/system/tinkernet-collator.service`

Within that file, paste in the following:
```
[Unit]
Description=Tinkernet (InvArch) Collator
After=network-online.target
Wants=network-online.target

[Service]
Type=simple
User=tinkernet-collator
Group=tinkernet-collator
ExecStart=/usr/local/bin/tinkernet-collator \
  --base-path /var/lib/tinkernet \
  --collator \
  --force-authoring \
  --name "YOUR-COLLATOR-NAME-HERE" \
  --chain /var/lib/tinkernet/tinker-raw.json \

>   --port 30333 \

Reviewer: ```suggestion
  --port 30333 \
  --listen-addr "/ip4/0.0.0.0/tcp/30333/ws" \
```
This makes the collator listen on websocket, future proofing to make interaction with light clients easier

Note: see more here if you want to double check the above https://docs.substrate.io/tutorials/get-started/trusted-network/#add-keys-to-the-keystore

Now we can start the the service again:

``sudo systemctl start tinkernet-collator.service``

Now we need to rotate keys and set our keys on chain.

Ensure the collator is running, or this step won't work:

``curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "author_rotateKeys", "params":[]}' http://127.0.0.1:9933/``

You will be greeted with an output that looks like:

``{"jsonrpc":"2.0","result":"0xprivate_key_will_be_here0","id":1}``

"result":"**0x_private_key_will_be_here0**" is what we are interested in.

You need to make sure that you have a Polkadot/Substrate account set up, here's some videos in case you don't know how to do that:

1. Polkadot JS Video  https://www.youtube.com/watch?v=dG0DP9vayPY    https://www.youtube.com/watch?v=BpTQBAyFvEk

2. Talisman Video   https://docs.talisman.xyz/talisman/talisman-initiation/setup-a-talisman-wallet  

> Now that you have made an account using one of those extensions, head on over to the GM Parachain section of Polkadot JS: 
> 
> https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fkusama.gmordie.com#/extrinsics

Reviewer: ```suggestion
Now that you have made an account using one of those extensions, head on over to the InvArch Tinkernet section of Polkadot JS: 

https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Ftinker.invarch.network#/extrinsics
```

> Ensure that you are in the Developer pallet (the top header), and navigate to the extrinsics section in the drop down.

Reviewer: ```suggestion
Ensure that you are in the Developer tab (the top header), and navigate to the extrinsics section in the drop down.
```

> Note: If you are doing this before the token distribution event, please ping the GM Intern in our Discord, and I will send you some $FREN so you can send the extrinsics.

Reviewer: ```suggestion
```
I'm afraid the GM Intern doesn't work here...

In the "using the selected account field" select the account you just made for the collator.<br/>
In the "submit the following extrinsic field" select "session".<br/>
In the next field (to the right), select "setKeys(keys, proof)".<br/>
In the "keys:" field, paste in your **0x_private_key_will_be_here0** from your node.<br/>
In the "proof" field, type in " 0 ".

Submit the transaction.

> Congratulations, you should now be onboarded as a collator.

Reviewer: Missing steps to register as a candidate.

Author: ✅ done

# How to set up an InvArch or Tinkernet Collator

Make sure you are logged in as a user with root privileges.

Create a new service user to run your collator service:
```shell
sudo useradd --no-create-home --shell /usr/sbin/nologin tinkernet-collator
```

You can either install the InvArch node using a pre-built binary from GitHub, or compile it from source if you'd like to customize it.

## Standard Installation

Download the pre-built binary to `/usr/local/bin` and give it the necessary permissions & ownership:
```shell
sudo wget https://github.com/InvArch/InvArch-Node/releases/latest/download/invarch-collator -P /usr/local/bin
sudo chmod +x /usr/local/bin/invarch-collator
sudo chown tinkernet-collator:tinkernet-collator /usr/local/bin/invarch-collator
```

## Install from Source

Install some required stuff:

```shell
curl https://sh.rustup.rs -sSf | sh
```
(choose option 1 - Proceed with installation (default))
You may need to restart your system before the next steps.
```

> sudo apt -y install cmake git clang libclang-dev

Reviewer: That's great

```

Clone the repo:

 ```shell

>  git clone git@github.com:InvArch/InvArch-Node.git

Author: ahhhh so it was just my user space fork that was requiring SSH key. good to know as I train others.

```

## Run the node

Now that you've got a collator node executable at `/usr/local/bin/tinkernet-collator` (either pre-built or built yourself), you're ready to create its data directory and run it for the first time.

Download the chainspec (`tinker-raw.json`), set up the `tinkernet` data directory, and give it the necessary ownership:
```shell
sudo mkdir /var/lib/tinkernet
sudo wget https://github.com/InvArch/InvArch-Node/releases/latest/download/tinker-raw.json -P /var/lib/tinkernet
sudo chown -R tinkernet-collator:tinkernet-collator /var/lib/tinkernet
```

Create a systemd service file to run your collator (and automatically restart it):

```shell
sudo nano /etc/systemd/system/tinkernet-collator.service
```

Within that file, paste in the following:
```
[Unit]
Description=Tinkernet (InvArch) Collator
After=network-online.target
Wants=network-online.target

[Service]
Type=simple
User=tinkernet-collator
Group=tinkernet-collator
ExecStart=/usr/local/bin/tinkernet-collator \
  --base-path /var/lib/tinkernet \
  --collator \
  --force-authoring \
  --name "YOUR-COLLATOR-NAME-HERE" \
  --chain /var/lib/tinkernet/tinker-raw.json \
  --listen-addr "/ip4/0.0.0.0/tcp/30333/ws" \

>   --telemetry-url "wss://telemetry.polkadot.io/submit 0" \

Author: reminder: you said you'd add bootnode to the chainspec

### Get the current candidate count

1. Visit the Regular UI: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Ftinker.invarch.network#/js
2. Add this script and run it:
```jsx
// Simple script to get candidate pool size
const candidatePool = await api.query.parachainStaking.candidatePool();
console.log(`Candidate pool size is: ${candidatePool.length}`);
```
3. Write down the result

### Join Candidates
1. Return to the Extrinics page: https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Ftinker.invarch.network#/extrinsics
2. In the `using the selected account` field, select the account you just made for the collator.
3. In the `submit the following extrinsic` field, select `collatorSelection`.
4. In the next field (to the right), select `registerAsCandidate()`.
5. Submit the transaction.
6. Wait 2 rounds (roughly 12 hours) for your collator registration to take effect.

Congratulations, you should now be onboarded as a collator.

## Prologue: Node Monitoring

As a collator node operator, you should also set up monitoring for your node, including Prometheus, Grafana, and AlertManager. Good instructions for that setup can already be found at:

> Prometheus/Grafana (HDX): https://docs.hydradx.io/node_monitoring

Author: hope it's ok to reference other chains. Better to not duplicate those instructions and have to maintain them.

Reviewer: InvArch 🤝 HydraDX
